src/fetch/affiliate_data.py

```diff
@@ -0,0 +1,126 @@
+    totalReferrals: int
+    referralVolumeUsd: float
+    lastUpdated: Optional[datetime]
+
+    @classmethod
+    def from_dict(cls, data_dict: DuneRecord, last_update: datetime) -> AffiliateData:
+        """Constructor of data class from DuneRecord or any dict with compatible types"""
+        return cls(
+            totalTrades=int(data_dict["total_trades"]),
+            totalReferrals=int(data_dict["total_referrals"]),
+            tradeVolumeUsd=float(data_dict["trade_volume_usd"]),
+            referralVolumeUsd=float(data_dict["referral_volume_usd"]),
+            lastUpdated=last_update,
+        )
+
+    @classmethod
+    def default(cls) -> AffiliateData:
+        """Default return type for NotFound affiliates"""
+        return cls(
+            totalTrades=0,
+            totalReferrals=0,
+            tradeVolumeUsd=0.0,
+            referralVolumeUsd=0.0,
+            lastUpdated=None,
+        )
+
+
+@dataclass
+class AffiliateMemory:
+    """Indexed affiliate data with knowledge of data's age"""
+
+    data: dict[str, AffiliateData]
+    last_update: datetime
+    last_execution: str
+
+
+class CachingAffiliateFetcher:
+    """
+    Class containing, DuneClient, FileIO and a logger for convenient Dune Fetching.
+    """
+
+    def __init__(
+        self, api_key: str, execution_id: Optional[str], cache_validity: int = 30
+    ) -> None:
+        """
+        Class constructor.
+        Builds DuneClient from `api_key`.
+        - `cache_validity` is in minutes.
+        - `execution_id` can be used to avoid initial fetching if known
+        """
+        log.info("loading affiliate data cache")
+        self.dune = DuneClient(api_key)
+        self.memory: AffiliateMemory = self._fetch_affiliates(execution_id)
+        self.cache_validity = timedelta(minutes=cache_validity)
+        log.info("affiliate data cache loaded")
+
+    def _fetch_affiliates(self, execution_id: Optional[str] = None) -> AffiliateMemory:
+        """Fetches Affiliate data from Dune and updates cache"""
+        if execution_id:
+            results = self.dune.get_result(execution_id)
+        else:
+            results = self.dune.refresh(Query(AFFILIATE_QUERY_ID), ping_frequency=10)
+
+        last_update = results.times.execution_ended_at or utc_now()
+        return AffiliateMemory(
+            last_update=last_update,
+            last_execution=results.execution_id,
+            data={
+                row["trader"]: AffiliateData.from_dict(row, last_update=last_update)
+                for row in results.get_rows()
+            },
+        )
+
+    def _update_memory(self) -> None:
+        """Internal method to update the cached affiliate data"""
+        self.memory = self._fetch_affiliates()
+
+    def cache_expired(self) -> bool:
+        """
+        Determines if data cache is expired.
+        """
+        before = self.memory.last_update
+        delta = self.cache_validity
+        return utc_now() > before + delta
```

Comment:
Just a discussion for the future since this is a bit late: In my opinion, it's okay for things like these, e.g., cache management, to depend on external packages, i.e., https://github.com/tkem/cachetools

Reply:
Yea it would be, except that while looking at some of these libraries I noticed that our logic was "too specialized" to use a generic decorator. This way, I feel like we have more specific control. I can look more into it if you think it would serve our purpose out of the box.

Reply:
> I can look more into it if you think it would serve our purpose out of the box.

No need. Just wanted to discuss. Thanks.